### ISSUE TITLE
Provide a way to set default Workflow options that will apply to all workflows

### ISSUE THREAD

**Issue by simster7** [argoproj/argo-workflows#1923]
# Summary

Provide a way to set default Workflow options that will apply to all workflows.

# Motivation

Users have asked for a way to set particular Workflow options once in the `worfklow-controller-cm` that will apply to all new Workflows as a convenience feature.

See: https://github.com/argoproj/argo/issues/1827#issuecomment-572126407 https://github.com/argoproj/argo/issues/1790 

# Proposal

We should add a way to allow users to set an arbitrary number of default Workflow options.

**Comment by NikeNano** [comment]
I would be happy to make a contribution and continue to work on this. I have looked over the code base a bit but don't understand how https://github.com/argoproj/argo/blob/master/docs/workflow-controller-configmap.yaml is parsed.

**Comment by rushtehrani** [comment]
We also need this for [PodGC](https://github.com/argoproj/argo/blob/02022e4bb36977253ff2e362f844b9596e768102/pkg/apis/workflow/v1alpha1/workflow_types.go#L234).

**Comment by NikeNano** [comment]
I have read more of the code base, and as I understand the [workflow-controller-configmap](https://github.com/argoproj/argo/blob/master/docs/workflow-controller-configmap.yaml) is parsed by the [config_controller](https://github.com/argoproj/argo/blob/master/workflow/controller/config_controller.go).  As I understand the workflowControllerConfig is parsed here: 
https://github.com/argoproj/argo/blob/02022e4bb36977253ff2e362f844b9596e768102/workflow/controller/config_controller.go#L39

I see the following updates needed: 
- Updated the workflow-controller-configmap to include the options of interest
- Pass the workflow-controller-configmap to the controllers of interest such as the [ttlcontroller](https://github.com/argoproj/argo/blob/master/workflow/ttlcontroller/ttlcontroller.go)
- Update the controllers of interest to handle the workflow-controller-configmap and use the relevant options.

**Comment by NikeNano** [comment]
@simster7  do you have some feedback on this? :)

**Comment by simster7** [comment]
We would have to go with option 1. However, we would also need to come up with a spec of how to declare default options on the config map. Do you have any ideas/suggestions for this?

**Comment by NikeNano** [comment]
Cool, thanks for the feedback @simster7. 

- Why option 1? It feels like it should be more general if we updated the workflow in one place instead? Then also it could be handled for other default settings instead of adding in the controllers? 

- For the config map I found a similar example for the https://github.com/argoproj/argo/blob/c51c1302f48cec5b9c6009b9b7e50962d338c679/docs/workflow-controller-configmap.yaml#L40

I suggest we add the following: 

```yaml 
 # TTLStrategy defines the default time to live for workflows depending on how they complete
 TTLStrategyDefault: 
    secondsAfterCompleted: -1 #  Default time to live after workflow is completed, replaces ttlSecondsAfterFinished
    secondsAfterSuccess: -1    #  Default time to live after workflow is successful
    secondsAfterFailed: -1   # Default time to live after workflow fails

```

I am not sure if we should add secondsAfterCompleted since it will be deprecated. 

What are your thoughts?

**Comment by NikeNano** [comment]
I it is fine that I start to work on it, could you please assign me to the issue?

**Comment by simster7** [comment]
@NikeNano Sorry, we've been rather busy with the upcoming 2.5 RC launch. I feel like we will need to have an internal discussion about this feature, which is not likely to happen until we ship the 2.5 RC launch.

However, feel free to make a proof of concept for this however you'd think it's best... if it's good it might drive our discussion in its favor. Just know that if you get started on this we might ask you to change it in the future :)

I will assign you the issue.

**Comment by zhujl1991** [comment]
> We would have to go with option 1. However, we would also need to come up with a spec of how to declare default options on the config map. Do you have any ideas/suggestions for this?

@simster7 Why do we go with option 1 here? Agree with the @NikeNano 's comment below:
"Why option 1? It feels like it should be more general if we updated the workflow in one place instead? Then also it could be handled for other default settings instead of adding in the controllers?"

**Comment by simster7** [comment]
Fixed in #2331
